Sources/Fuzzilli/FuzzIL/TypeSystem.swift

```diff
@@ -850,53 +789,155 @@ class TypeExtension: Hashable {
+            case .rest(_):
+                return true
+            default:
+                return false
+        }
+    }
+
+    var isOptional: Bool {
+        switch self {
+            case .opt(_):
+                return true
+            default:
+                return false
+        }
+    }
+
+    public typealias ProtobufType = Fuzzilli_Protobuf_Parameter
+
+    func asProtobuf(with typeCache: TypeCache?) -> ProtobufType {
+        switch self {
+            case .plain(let inputType):
+                return ProtobufType.with {
+                    $0.plainParameter = Fuzzilli_Protobuf_PlainParameter.with {
+                        $0.inputType = inputType.asProtobuf(with: typeCache)
+                    }
+                }
+            case .opt(let inputType):
+                return ProtobufType.with {
+                    $0.optionalParameter = Fuzzilli_Protobuf_OptionalParameter.with {
+                        $0.inputType = inputType.asProtobuf(with: typeCache)
+                    }
+                }
+            case .rest(let inputType):
+                return ProtobufType.with {
+                    $0.restParameter = Fuzzilli_Protobuf_RestParameter.with {
+                        $0.inputType = inputType.asProtobuf(with: typeCache)
+                    }
+                }
+        }
+    }
+
+    public func asProtobuf() -> ProtobufType {
+        return asProtobuf(with: nil)
+    }
+
+    init(from proto: ProtobufType, with typeCache: TypeCache?) throws {
+        switch proto.param {
+            case .plainParameter(let p):
+                self = .plain(try Type(from: p.inputType, with: typeCache))
+            case .optionalParameter(let p):
+                self = .opt(try Type(from: p.inputType, with: typeCache))
+            case .restParameter(let p):
+                self = .rest(try Type(from: p.inputType, with: typeCache))
+            default:
+              throw FuzzilliError.typeDecodingError("invalid parameter type")  
+        }
+    }
+
+    public init(from proto: ProtobufType) throws {
+        try self.init(from: proto, with: nil)
+    }
+}
+
 public struct FunctionSignature: Hashable, CustomStringConvertible {
-    public let inputTypes: [Type]
+    public let parameters: [Parameter]
     public let outputType: Type
 
     public func format(abbreviate: Bool) -> String {
-        let params = inputTypes.map({ $0.format(abbreviate: abbreviate) }).joined(separator: ", ")
+        let params = parameters.map({ $0.format(abbreviate: abbreviate) }).joined(separator: ", ")
         return "[\(params)] => \(outputType.format(abbreviate: abbreviate))"
     }
 
     public var description: String {
         return format(abbreviate: false)
     }
 
-    public init(expects parameterTypes: [Type], returns returnType: Type) {
-        self.inputTypes = parameterTypes
+    // Returns the number of innerOutputs that are generated from the function signature
+    public var numParameters: Int {
```

Comment:
Taking into account future PRs about parameter destructuring, is this really true? I guess for something like
`function foo(a, [b, c, d], e))` we'd have `numParameters = 3` (caller side) but `numInnerOutputs (or numCreatedVariables) = 5` (callee side). So I think we should clarify whether these counts are valid for the caller or the callee :)

Reply:
So the direction I was going with this was that `numParameters` would represent callee side variables. In the example above `numParameters` would be 5 and to calculate the number of caller side arguments (which is what's being done in ProgramBuilder) I use the count from `callerTypes`.
Happy to change this though if that would help simplify this.

Reply:
Ok, yeah WDYT about renaming it to something like `numOutputVariablesInCallee`? Maybe a bit verbose, but at least it's fairly obvious (I think) what it expresses. I think it'll be confusing to have `sig.numParameters != sig.parameters.count` in the future... :)